pkg/autoscaler/metrics/stats_scraper.go

```diff
@@ -181,7 +189,111 @@ func (s *serviceScraper) Scrape(window time.Duration) (Stat, error) {
 	if readyPodsCount == 0 {
 		return emptyStat, nil
 	}
-	return s.scrapeService(window, readyPodsCount)
+
+	startTime := time.Now()
+	scrapeTime := time.Since(startTime)
+	pkgmetrics.RecordBatch(s.statsCtx, scrapeTimeM.M(float64(scrapeTime.Milliseconds())))
+
+	if s.podsAddressable {
+		stat, err := s.scrapePods(readyPodsCount)
+		// Some pods were scraped, but not enough.
+		if err != errNoPodsScraped {
+			return stat, err
+		}
+		// Else fall back to service scrape.
+	}
+	stat, err := s.scrapeService(window, readyPodsCount)
+	if err == nil {
+		// If err == nil, this means that we failed to scrape all pods, but service worked
+		// thus it is probably a mesh case.
+		s.podsAddressable = false
```

Comment:
Are users able to go back to direct pod scraping after disabling mesh? Possibly restarting autoscaler?

Reply:
Yes and you have to do that anyway

Reply:
The detection is per revision, so a new revision without the mesh would be able to take advantage of pod addressability. It also allows a hybrid approach where some revisions are in a mesh and some are not. That‘s the same in the activator too.